### PR TITLE
Replace nix with poetry

### PR DESCRIPTION
**Pull request by slovdahl**
Thanks to the original author @qqshfox in https://github.com/smarkets/marge-bot/pull/291. Will keep it in our fork until it has hopefully been merged upstream.

### CODE REVIEW

**Comment by perlun** [comment]
Hmm... slightly painful CI failures on this one. :thinking: Could there be some missing dependency somewhere?

```
        gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Iast27/Include -I/home/runner/.cache/pypoetry/virtualenvs/marge-YHqeRpzV-py3.10/include -I/opt/hostedtoolcache/Python/3.10.9/x64/include/python3.10 -c ast27/Custom/typed_ast.c -o build/temp.linux-x86_64-cpython-310/ast27/Custom/typed_ast.o
        In file included from ast27/Custom/../Include/Python-ast.h:3,
                         from ast27/Custom/typed_ast.c:2:
        ast27/Custom/../Include/../Include/asdl.h:32:41: error: unknown type name ‘PyArena’
           32 | asdl_seq *asdl_seq_new(Py_ssize_t size, PyArena *arena);
              |                                         ^~~~~~~
        ast27/Custom/../Include/../Include/asdl.h:33:49: error: unknown type name ‘PyArena’
           33 | asdl_int_seq *asdl_int_seq_new(Py_ssize_t size, PyArena *arena);
              |                                                 ^~~~~~~
```

**Comment by perlun** [comment]
This one seems to have some info: https://github.com/python/typed_ast/issues/159

**Comment by slovdahl** [comment]
Hmm.. The requirement is defined like this:
```
typed-ast==1.4.2; implementation_name == "cpython" and python_version < "3.8" and python_version >= "3.6" \
```

So how do we even update it for Python 3.10? :thinking:

**Comment by perlun** [comment]
TL;DR: python/typed_ast (whatever the package name happens to be) should be upgraded to at least 1.4.3, to support Python 3.10.

**Comment by perlun** [comment]
> So how do we even update it for Python 3.10? thinking

Hmm, that makes it more complex then yes... :thinking: Will dig a bit more.

**Comment by slovdahl** [comment]
My guess is that it's a transitive dependency of some sort.. I think one problem is that pytest was downgraded. I'm changing that back to 7.0.1 now with Poetry.

**Comment by perlun** [comment]
> Hmm.. The requirement is defined like this:

Where is this? :thinking: I can't seem to find any references to this in this repo, apart from in `requirements.nix`:

```
$ rg typed-ast
requirements.nix
118:      self."typed-ast"
685:    "typed-ast" = python.mkDerivation {
686:      name = "typed-ast-1.3.5";
687:      src = pkgs.fetchurl { url = "https://files.pythonhosted.org/packages/d3/b1/959c3ed4a9cc100feba7ad1a7d6336d8888937ee89f4a577f7698e09decd/typed-ast-1.3.5.tar.gz"; sha256 = "5315f4509c1476718a4825f45a203b82d7fdf2a6f5f0c8f166435975b1c9f7d4"; };
```

**Comment by slovdahl** [comment]
Better, that seems to have been the problem :+1:

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+[tool.poetry]
+name = "marge"
+version = "0.10.0"
+description = "Marge-bot is a merge-bot for GitLab that, beside other goodies, implements [the Not Rocket Science Rule Of Software Engineering](http://graydon2.dreamwidth.org/1597.html)."
+license = "BSD-3-Clause"
+authors = ["Infrastructure <infrastructure@smarkets.com>"]
+readme = "README.md"
+repository = "https://github.com/smarkets/marge-bot"
```
Comment (slovdahl) [inline pyproject.toml]: Fix this before merging.

File: README.md [hunk 2]
```diff
@@ -364,19 +364,8 @@ run:
...
-The nix install should be fully reproducible on any version of linux (and also
-work on OS X, although this is not something we properly test). If you don't
-want to use docker we recommend you give nix a try.
-
-#### Installing marge-bot the old-fashioned way
-
-Finally, although this is our least preferred alternative, you can always do
-`python3 setup.py install` (note that you will need python3.6).
+You need to install the requirements first `pip install -r requirements_frozen.txt`,
+and then install Marge itself `python3 setup.py install` (note that you will need python3.6+).
```
Comment (slovdahl) [inline README.md]: Fix the development docs further down in this file too.

**Comment by slovdahl** [comment]
Tested that building a Docker image using `make dockerize` works too.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -12,7 +12,7 @@ env:
   PY_COLORS: 1
 
 jobs:
-  test:
+  pytest:
```
Comment (perlun) [inline .github/workflows/test.yml]: We'll need to remember to adjust the "required jobs to succeed" setting after merging this I think.

Comment (perlun) [reply .github/workflows/test.yml]: I.e. these. So perhaps it would be simpler to just keep it as `test` or WDYT?

image

Comment (slovdahl) [reply .github/workflows/test.yml]: Hmm, where do you find these settings? But sure, that works for me. That also explains why these `test` jobs show up as "required" here..

Comment (slovdahl) [reply .github/workflows/test.yml]: Ah, they need to be changed regardless, because I introduce a new variable.

Comment (perlun) [reply .github/workflows/test.yml]: It's under the branch protection settings for the `master` branch: https://github.com/hiboxsystems/marge-bot/settings/branch_protection_rules/32450687

**Review by perlun — APPROVED** [review]

File: .github/workflows/test.yml (hunk 3 above)
Comment (slovdahl) [reply .github/workflows/test.yml]: Hm, ok. Can we just require the pipeline to be green and not require specific jobs? :thinking:

Comment (slovdahl) [reply .github/workflows/test.yml]: I removed the specific checks now.

Comment (perlun) [reply .github/workflows/test.yml]: > Can we just require the pipeline to be green and not require specific jobs? thinking

I'm not sure that works in fact. Try provoking an error now, I think GitHub will still let us merge if we configure it like that. :unamused:

Comment (slovdahl) [reply .github/workflows/test.yml]: image

Comment (slovdahl) [reply .github/workflows/test.yml]: image

Comment (slovdahl) [reply .github/workflows/test.yml]: What insanity is this :grin: But is it a huge problem? It's not green and shiny as it is when all is green, so maybe we can live with it?

Comment (perlun) [reply .github/workflows/test.yml]: Insanity it is; not all git providers can apparently be as convenient as GitLab in this regard. :wink: But yes, let's live with it for now. :+1:
